### PR TITLE
docs: expand description for --dns.resolvers

### PR DESCRIPTION
**Pull request by dmke**
This extends the documentation for the `--dns.resolvers` flag as suggested in #1777.

The commits are currently based off of PR #1785, so depending on whether that gets merged, I'll rebase this PR accordingly.

@oseiberts11: Would you like to proof-read the new section? You can find a preview here: https://mm.dmke.org/lego-b649657b/usage/cli/options/#dns-resolvers-and-challenge-verification

### CODE REVIEW

**Comment by oseiberts11** [comment]
@dmke Thanks! This looks great to me.

File: cmd/flags.go [hunk 1]
```diff
@@ -102,8 +103,11 @@ func CreateFlags(defaultPath string) []cli.Flag {
 			Usage: "By setting this flag to true, disables the need to wait the propagation of the TXT record to all authoritative name servers.",
 		},
 		&cli.StringSliceFlag{
-			Name:  "dns.resolvers",
-			Usage: "Set the resolvers to use for performing recursive DNS queries. Supported: host:port. The default is to use the system resolvers, or Google's DNS resolvers if the system's cannot be determined.",
+			Name: "dns.resolvers",
+			Usage: "Set the resolvers to use for performing (recursive) CNAME resolving and apex domain determination." +
+				" For DNS challenge verification, the authoritative DNS server is queried directly." +
+				" Supported: host:port." +
+				" The default is to use the system resolvers, or Google's DNS resolvers if the system's cannot be determined.",
```
Comment (jsumners) [inline cmd/flags.go:110]: Please also include an example. It's not clear how to provide multiple values for this setting. Do you provide a CSV or something else?

Comment (jsumners) [inline cmd/flags.go:110]: I'm still not clear what this means. What does setting this value gain a user?

Comment (dmke) [reply cmd/flags.go:110]: (I'll need to provide a bit of context, please bear with me for a moment. I'll try to be brief, but that also requires to simplify things. This is the ELI5 version.)

---

The ACME protocol involves a few steps of verification, in order to issue you a TLS certificate. One of the earliest steps is to ask the ACME provider (usually Let's Encrypt), if `lego` can have such a certificate for `www.example.com`, to which Let's Encrypt answers, "sure, if you can prove you own that domain; here is a token and place it someplace only accessible to owners".

Proving ownership can happen in different ways, but the bulk of the code in this repository is concerned with the "DNS challenge" (dns-01). This means, `lego` must place the aforementioned token as TXT record on `_acme-challenge.www.example.com`.

Now, DNS being the redundant and resilient system it is, is not known for fast propagation of changes. Layers of caching servers are setup between you and the name server responsible to deliver you that TXT record (on your PC, your router, your ISP have certainly installed a recursive resolver). Each layer may (or may not!) obey the caching time for certain DNS answers, even for negative answers ("I don't know about `_acme-challenge.www.example.com`, and the authoritative name server has told me to save this answer for 600 seconds").

This makes the verifying step tedious. `lego` could skip this step, and leave it to Let's Encrypt. But should setting up the TXT record fail (or just take a moment), Let's Encrypt may just respond with "you didn't setup the token, you don't get a certificate", and deduct one point from your request quota (remember that Let's Encrypt rate limits your account to 300 new orders per 3 hours). If you try and retry and retry to obtain a certificate, your quota may be exhausted quicker than you think.

The next less naïve approach would be to wait a moment. But that begs the question: for how long? Should every user generally suffer from overly long waiting periods, just because in some cases the propagation takes a bit longer? Certainly not.

So... what does `lego` do? Easy: We're querying the authoritative DNS server directly, avoiding any caches. The not-so-easy part is determining which server that is.

For a domain like `example.com`, we can just lookup the SOA record for `example.com`, and extract the MNAME from it (in DNS lingo, `example.com` is the apex domain for `example.com`, because it contains the SOA record). This however gets a little more complicated for domains like `zeus.fra.ger.eu.cdn.example.com`, where the whole `*.cdn.example.com` zone might be delegated to another server (the apex domain might be `cdn.example.com`, not `example.com`).

Here, `lego` can only ask first `zeus.fra.ger.eu.cdn.example.com` for its SOA record, then `fra.ger.eu.cdn.example.com`, etc., until it finds the authoritative name server.

The final step is to query that name server, "hey what's the TXT record for `_acme-challenge.zeus.fra.ger.eu.cdn.example.com`? If it doesn't provide the correct answer, `lego` will wait a moment (`--dns-timeout`) and try again (I forgot how often).

(The story then continues with a similar verification on Let's Encrypt's side, and the exchange of the TLS certificate, should LE be satisfied.)

---

What does setting [the `--dns.resolvers`] value gain a user?

Since `lego`, by default, uses the system's DNS resolver to determine the apex domain, this may lead to a wrong answer. Within data centers you might have DNS setup in a way which gives you a different answer than queries in from the outside world ("split DNS", "hidden master").

A false-positive here is not better than the first approach described above (i.e. let Let's Encrypt figure it out). A false-negative here could mean that your system resolver is not able at all to get the `_acme-challenge` record - and `lego` would wait and wait and finally fail.

Another reason can be that you don't trust your system's/router's/ISP's name server. I know from some residential ISPs here in Germany where in the past they wanted to be helpful and present users with a "Did you mean alice-dsl.de? Here are some ads"-page, if the end-user had a typo when entering a domain into the browsers URL bar (/rant). Of course such a behaviour breaks `lego`'s verification process as well.

In such situations, you need to instruct `lego` to use a specific (set of) known-good resolver(s) (ideally but not necessarily on the public internet). The `--dns.resolvers` flag is for this specific requirement.

Comment (dmke) [reply cmd/flags.go:110]: If you run `lego --help` (or see [here](https://github.com/go-acme/lego/pull/1786/files#diff-fe0fb6c80906c79ff932549d18a30a1cff5846e1ac80a7d64a30baa965a0d74fR28), you'll find that you can just specify the `--dns.resolvers` flag multiple times:

```
$ lego --help

...

   --dns.resolvers value [ --dns.resolvers value ]
              Set the resolvers to use for performing (recursive) CNAME resolving and
              apex domain determination. For DNS challenge verification, the authoritative
              DNS server is queried directly. Supported: host:port. The default is to use
              the system resolvers, or Google's DNS resolvers if the system's cannot be
              determined.
```

(I've reformatted it a bit to cope with the restricted space imposed by Github.)

Comment (jsumners) [reply cmd/flags.go:110]: Is that new? I have v4.8.0 (via https://github.com/void-linux/void-packages/blob/master/srcpkgs/lego/template) and I see:

```
> lego --help
**snip**
--dns.resolvers value        Set the resolvers to use for performing recursive DNS queries. Supported: host:port. The default is to use the system resolvers, or Google's DNS resolvers if the system's cannot be determined.
**snip**
```

Still, I think it would be good to state that the option can be repeated multiple times for multiple resolvers. This is the first time I can recall seeing this bracket syntax and it isn't super clear.

Comment (dmke) [reply cmd/flags.go:110]: Hm, curious. I'll have to investigate...

Comment (dmke) [reply cmd/flags.go:110]: Looks like we've updated the CLI argument parser library in Lego 4.9.0 from [2.3.0 to 2.14.0](https://github.com/urfave/cli/compare/v2.3.0...v2.14.0), which could explain differences in the output.

Comment (jsumners) [reply cmd/flags.go:110]: 👍

File: docs/content/usage/cli/Options.md [hunk 2]
```diff
@@ -131,3 +47,23 @@ If you are using this option, make sure you proxy all of the following traffic t
...
+   It contains the SOA record, from which Lego extracts the authoritative name server (MNAME, i.e. the primary name server).
+   Should any DNS label on the way be a CNAME, it is resolved as per usual.
+
+   To resolve the apex domain and CNAMEs, Lego sends queries to the configured DNS resolvers.
+   These are, by default, the system name servers, and fallback to Google's DNS servers, should they be absent. You can override this behaviour with the `--dns.resolvers` flag.
+
+2. Verifying the challenge token.
+
+   The `_acme-challenge.<yourdomain>` TXT record must be correctly installed.
+   Lego verifies this by directly querying the authoritative name server for this record (as detected in the previous step).
```
Comment (jsumners) [inline docs/content/usage/cli/Options.md:69]: image

**Comment by dmke** [comment]
> @ldez marked this pull request as ready for review 2 minutes ago

Github doesn't let me approve my own PR, but the changes LGTM.

**Comment by ldez** [comment]
I just rebased the PR, there are no changes :wink:

File: docs/content/usage/cli/Options.md [hunk 3]
```diff
@@ -47,3 +47,23 @@ If you are using this option, make sure you proxy all of the following traffic t
...
 
 [^header]: You must ensure that incoming validation requests contains the correct value for the HTTP `Host` header. If you operate lego behind a non-transparent reverse proxy (such as Apache or NGINX), you might need to alter the header field using `--http.proxy-header X-Forwarded-Host`.
+
+## DNS Resolvers and Challenge Verification
+
+When using a DNS challenge provider (via `--dns <name>`), Lego tries to ensure the ACME challenge token is properly setup before instructing the ACME provider to perform the validation.
+
+This involves a few DNS queries to different servers:
+
+1. Determining the apex domain and resolving CNAMES.
```
Comment (ldez) [inline docs/content/usage/cli/Options.md:57]: The goal is to find the "zone", it's not really the apex, it can be the 3rd, 4th, 5th, 6th, etc. level

**Comment by dmke** [comment]
> I just rebased the PR, there are no changes :wink:

Sorry, I've been looking at the changes of #1785.

File: docs/content/usage/cli/Options.md (hunk 3 above)
Comment (dmke) [reply docs/content/usage/cli/Options.md:57]: Isn't the apex domain the domain containing the SOA record?

Comment (ldez) [reply docs/content/usage/cli/Options.md:57]: Depend on the definition of the apex.
We try to find the domain that handles the subdomain, it can be a subdomain.
An SOA record can be placed on a subdomain.
For me, the apex is the ltd + the "first domain", but maybe my own definition is wrong.

Comment (ldez) [reply docs/content/usage/cli/Options.md:57]: I hadn't seen your edit.

Comment (dmke) [reply docs/content/usage/cli/Options.md]: I've reworded the section and added a bit more content ([preview](https://mm.dmke.org/lego-b649657b/usage/cli/options/#dns-resolvers-and-challenge-verification)).

**Review by ldez — APPROVED** [review]
LGTM
